File: webauthn.go
package main

import (
	"encoding/base64"
	"encoding/binary"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"strings"

	"github.com/go-webauthn/webauthn/protocol"
	"github.com/go-webauthn/webauthn/webauthn"
)

type WebAuthn struct {
	wa    *webauthn.WebAuthn
	store *Store

	// TODO: concurrency
	registrationSessions map[uint32]*webauthn.SessionData
	loginSessions        map[string]*webauthn.SessionData
}

func NewWebAuthn(store *Store, hostname string, displayName string, origins []string) *WebAuthn {
	config := &webauthn.Config{
		RPID:          hostname,
		RPDisplayName: displayName,
		RPOrigins:     origins,
	}
	wa, err := webauthn.New(config)
	if err != nil {
		panic(err)
	}
	return &WebAuthn{
		store: store,
		wa:    wa,

		registrationSessions: make(map[uint32]*webauthn.SessionData),
		loginSessions:        make(map[string]*webauthn.SessionData),
	}
}

func writeJsonBody(w http.ResponseWriter, data any) error {
	w.Header().Add(`Content-Type`, `application/json`)
	encoder := json.NewEncoder(w)
	if err := encoder.Encode(data); err != nil {
		log.Println(err)
		return err
	}
	return nil
}

func (a *WebAuthn) Handler(prefix string) http.Handler {
	mux := http.NewServeMux()

	mux.HandleFunc(`POST /register:begin`, a.BeginRegistration)
	mux.HandleFunc(`POST /register:finish`, a.FinishRegistration)
	mux.HandleFunc(`POST /login:begin`, a.BeginLogin)
	mux.HandleFunc(`POST /login:finish`, a.FinishLogin)

	// 垃圾 js，转换个 ArrayBuffer 和 base64 都麻烦得要死。
	mux.HandleFunc(`POST /base64:encode`, a.base64Encode)
	mux.HandleFunc(`POST /base64:decode`, a.base64Decode)

	return http.StripPrefix(strings.TrimSuffix(prefix, "/"), mux)
}

// [1,2,3] => XXXXXX
// 官方的 protocol.URLEncodedBase64 会把结果转成 JSON 字符串，不好用。
func (a *WebAuthn) base64Encode(w http.ResponseWriter, r *http.Request) {
	var bs []byte
	if err := json.NewDecoder(io.LimitReader(r.Body, 1<<20)).Decode(&bs); err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}
	str := base64.RawURLEncoding.EncodeToString(bs)
	w.Write([]byte(str))
}

// XXXXXX => [1,2,3]
func (a *WebAuthn) base64Decode(w http.ResponseWriter, r *http.Request) {
	var s string
	if all, err := io.ReadAll(io.LimitReader(r.Body, 1<<20)); err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	} else {
		s = string(all)
	}
	bs, err := base64.RawURLEncoding.DecodeString(s)
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}
	ibs := make([]int, 0, len(bs))
	for _, b := range bs {
		ibs = append(ibs, int(b))
	}
	if err := json.NewEncoder(w).Encode(ibs); err != nil {
		log.Println(err)
		return
	}
}

func (a *WebAuthn) BeginRegistration(w http.ResponseWriter, r *http.Request) {
	user := a.store.AuthRequest(r)
	if user == nil {
		http.Error(w, "You should be logged in before you can register a credential.", http.StatusForbidden)
		return
	}

	options, session, err := a.wa.BeginRegistration(user,
		// this is for discoverable login. Discoverable login do not require you
		// to first enter your username/email, it it's you select a user from
		// the list.
		webauthn.WithResidentKeyRequirement(protocol.ResidentKeyRequirementRequired),
	)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	if err := writeJsonBody(w, options); err != nil {
		return
	}

	a.registrationSessions[user.ID] = session
}

func (a *WebAuthn) FinishRegistration(w http.ResponseWriter, r *http.Request) {
	user := a.store.AuthRequest(r)
	if user == nil {
		panic(`login first`)
	}

	session, ok := a.registrationSessions[user.ID]
	if !ok {
		http.Error(w, "session not found", http.StatusBadRequest)
		return
	}
	delete(a.registrationSessions, user.ID)

	credential, err := a.wa.FinishRegistration(user, *session, r)
	if err != nil {
		http.Error(w, "Registration fail："+err.Error(), http.StatusInternalServerError)
		return
	}

	a.store.AddWebAuthnCredentialFor(user, credential)

	w.WriteHeader(http.StatusOK)
}

func (a *WebAuthn) BeginLogin(w http.ResponseWriter, r *http.Request) {
	options, session, err := a.wa.BeginDiscoverableLogin()
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	if err := writeJsonBody(w, options); err != nil {
		return
	}
	a.loginSessions[session.Challenge] = session
}

func (a *WebAuthn) FinishLogin(w http.ResponseWriter, r *http.Request) {
	challenge := r.URL.Query().Get(`challenge`)
	session, ok := a.loginSessions[challenge]
	if !ok {
		http.Error(w, "session not found", http.StatusBadRequest)
		return
	}
	delete(a.loginSessions, challenge)

	var user *User

	// if no error, we successfully logged in.
	credential, err := a.wa.FinishDiscoverableLogin(a.findUser(&user), *session, r)
	if err != nil {
		http.Error(w, err.Error(), http.StatusUnauthorized)
		return
	}

	// as per the lib, we should update credential list.
	a.store.AddWebAuthnCredentialFor(user, credential)

	// keep logged in from browser by setting cookie.
	a.store.MakeCookie(user, w, r)

	w.WriteHeader(http.StatusOK)
}

func (a *WebAuthn) findUser(user **User) func(rawID, userHandle []byte) (webauthn.User, error) {
	return func(rawID, userHandle []byte) (webauthn.User, error) {
		// just in case
		if *user != nil {
			return nil, fmt.Errorf(`user already found`)
		}
		if len(userHandle) != 4 {
			return nil, fmt.Errorf(`bad user handle length: %v`, len(userHandle))
		}
		id := binary.LittleEndian.Uint32(userHandle)
		u := a.store.GetUserByID(id)
		if u != nil {
			*user = u
			return u, nil
		}
		return nil, fmt.Errorf(`no such user`)
	}
}

File: webauthn.js
class WebAuthnAPI {
	constructor(){}

	async beginRegistration() {
		let path = `/admin/login/webauthn/register:begin`;
		let rsp = await fetch(path, {
			method: 'POST',
		});
		if (!rsp.ok) {
			throw new Error(await rsp.text());
		}
		let options = await rsp.json();
		options.publicKey.challenge = await this.base64decode(options.publicKey.challenge);
		options.publicKey.user.id = await this.base64decode(options.publicKey.user.id);
		return options;
	}
	async finishRegistration(options) {
		options.rawId = await this.base64encode(options.rawId);
		options.response.clientDataJSON = await this.base64encode(options.response.clientDataJSON);
		options.response.attestationObject = await this.base64encode(options.response.attestationObject);

		let path = `/admin/login/webauthn/register:finish`;
		let rsp = await fetch(path, {
			method: 'POST',
			body: JSON.stringify(options),
		});
		if (!rsp.ok) {
			throw new Error(await rsp.text());
		}
	}
	async beginLogin() {
		let path = `/admin/login/webauthn/login:begin`;
		let rsp = await fetch(path, {
			method: 'POST',
		});
		if (!rsp.ok) {
			throw new Error(await rsp.text());
		}
		let options = await rsp.json();
		options.publicKey.challenge = await this.base64decode(options.publicKey.challenge);
		return options;
	}
	// challenge: 只是用来查找 session 的，肯定不会用作服务端真实 challenge。想啥呢？
	async finishLogin(options, challenge) {
		let path = `/admin/login/webauthn/login:finish?challenge=${challenge}`;
		options.rawId = await this.base64encode(options.rawId);
		options.response.authenticatorData = await this.base64encode(options.response.authenticatorData);
		options.response.clientDataJSON = await this.base64encode(options.response.clientDataJSON);
		options.response.signature = await this.base64encode(options.response.signature);
		options.response.userHandle = await this.base64encode(options.response.userHandle);
		let rsp = await fetch(path, {
			method: 'POST',
			body: JSON.stringify(options),
		});
		if (!rsp.ok) {
			throw new Error(await rsp.text());
		}
	}

	async base64encode(arr) {
		let path = '/admin/login/webauthn/base64:encode';
		let body = JSON.stringify(Array.from(new Uint8Array(arr)));
		let rsp = await fetch(path, { method: 'POST', body: body});
		return await rsp.text();
	}
	async base64decode(str) {
		let path = '/admin/login/webauthn/base64:decode';
		let rsp = await fetch(path, { method: 'POST', body: str});
		return new Uint8Array(await rsp.json());
	}
}

class WebAuthn {
	constructor() {
		this.api = new WebAuthnAPI();
	}

	async register() {
		let { publicKey } = await this.api.beginRegistration();
		console.log('开始注册返回：', publicKey);

		let { rp, challenge, user } = publicKey;
		
		let createOptions = {
			challenge: challenge,
			rp: {
				name: rp.name,
				id: rp.id,
			},
			pubKeyCredParams: [
				{
					alg: -8,    // ed25519，我的最爱！❤️
					type: 'public-key',
				},
			],
			user: {
				id: user.id,
				name: user.name,
				displayName: user.displayName,
			},
		};
		console.log('创建参数：', createOptions);

		let credential = null;
		try {
			credential = await navigator.credentials.create({ publicKey: createOptions });
		} catch (e) {
			console.log(e);
			throw e;
		}
		if (!credential) { return null; }
		console.log('本地创建凭证：', credential);

		let finishOptions = {
			id:     credential.id,
			type:   credential.type,
			rawId:  credential.rawId,
			response: {
				clientDataJSON:     credential.response.clientDataJSON,
				attestationObject:  credential.response.attestationObject,
			},
		};
		console.log('结束注册参数：', finishOptions);

		await this.api.finishRegistration(finishOptions);
		console.log('注册成功。');
	}

	async login() {
		let { publicKey } = await this.api.beginLogin();
		console.log('开始登录返回：', publicKey);
		let options = {
			challenge: publicKey.challenge,
		};
		let credential = await navigator.credentials.get({ publicKey: options});
		console.log('本地获取凭证：', credential);

		let finishLoginOptions = {
			id:     credential.id,
			type:   credential.type,
			rawId:  credential.rawId,
			response: {
				clientDataJSON:     credential.response.clientDataJSON,
				authenticatorData:  credential.response.authenticatorData,
				signature:          credential.response.signature,
				userHandle:         credential.response.userHandle,
			},
		};
		console.log('结束登录参数：', finishLoginOptions);
		await this.api.finishLogin(finishLoginOptions, await this.api.base64encode(publicKey.challenge));
		console.log('登录成功。');
	}
}
